### PR TITLE
Fix crash when checking module exports for export=

### LINKED ISSUE #10460: getDeclarationOfAliasSymbol assert with shortid npm module in 2@next
**TypeScript Version:**  2.1.0 / nightly (2.1.0-dev.20160821)


**Error Message**
```
G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:37610                                                            
                throw e;                                                                                                            
                ^                                                                                                                   
                                                                                                                                    
Error: Debug Failure. False expression:                                                                                             
    at Object.assert (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:1289:23)                                   
    at Object.fail (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:1294:19)                                     
    at Object.findMap (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:211:15)                                   
    at getDeclarationOfAliasSymbol (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:13927:23)                    
    at checkExternalModuleExports (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:26428:39)                     
    at checkSourceFileWorker (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:26635:21)                          
    at checkSourceFile (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:26610:13)                                
    at Object.forEach (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:185:30)                                   
    at getDiagnosticsWorker (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:26659:16)                           
    at getDiagnostics (G:\Sean\Documents\Projects\test\node_modules\typescript\lib\tsc.js:26647:24)                                 
```

**package.json**

```json
{
  "name": "test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "typings": "typings",
    "tsc": "tsc"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "shortid": "^2.2.6",
    "typescript": "^2.1.0-dev.20160821",
    "typings": "^1.3.2"
  }
}
```

**tsconfig.json**
```json
{
    "compileOnSave": true,
    "compilerOptions": {
        "module": "commonjs",
        "moduleResolution": "node",
        "target": "es6",
        "noImplicitAny": true,
        "strictNullChecks": true,
        "sourceMap": true,
        "allowJs": true,
        "outDir": "bin"
    },
    "files": ["index.ts"]
}
```

**typings.json**
```json
{
  "globalDependencies": {
    "shortid": "registry:dt/shortid#0.0.0+20160316155526"
  }
}
```

**index.ts**
```ts
import * as shortid from 'shortid';
```

**Reproduction Steps**
```
npm install
npm run typings -- install
npm run tsc
```

**Note**

Removing the `Object.fail` in `Object.findMap` results in a crash-free compilation with no ill-effects that I can observe, including on a somewhat larger project.

`shortid` is not the only module that causes this. I've also observed `debug` causing this. I'm not sure how many others my app might be hitting since I can't see latter failures until somehow avoiding former ones.

### LINKED ISSUE #10522: Tsc crash ("Debug Failure. False expression:")
Hi,
I think the following code should not cause tsc to crash:

**TypeScript Version:**  Version 2.1.0-dev.20160824

**Code** (test1.ts)

```ts
export let a;
export = {};
```

**Expected behavior:**
```
test1.ts(2,1): error TS2309: An export assignment cannot be used in a module with other exported elements.
```

**Actual behavior:** 
```
C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:37652
                throw e;
                ^

Error: Debug Failure. False expression:
    at Object.assert (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:1289:23)
    at Object.fail (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:1294:19)
    at Object.findMap (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:211:15)
    at getDeclarationOfAliasSymbol (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:13935:23)
    at checkExternalModuleExports (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26465:39)
    at checkExportAssignment (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26441:13)
    at checkSourceElement (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26609:28)
    at Object.forEach (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:185:30)
    at checkSourceFileWorker (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26661:20)
    at checkSourceFile (C:\Users\VPC\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26647:13)
```

Note: The error seems to have been introduced with commit e900952d3d941524c2de373512693a8421eee22d (with c9f62f33d4d87dbb4a40ba5a28bb1085c683def2 I'm getting the expected error message).

Thanks!

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #10460 and fixes #10522 

This PR also makes maxNodeModuleJsDepth default to 0 so that incorrect tsconfigs now let the compiler spend less time compiling JS that is found in node_modules (especially since most people will already have the d.ts and want ignore the JS anyway). 

jsconfig still defaults to 2.

### CODE REVIEW

**Comment by sandersn** [comment]
@andy-ms and @vladima can you take a look?

**Comment by sandersn** [comment]
A baselining test that asserts that `allowJs: true` + `"exclude": [ "typings" ]` produces the node_modules-js-compilation behaviour from #10460 is blocked by a bug in the test harness (#10539). And I think it's a pretty low-value test anyway.

**Comment by sandersn** [comment]
@mhegazy you should also know about this fix.

File: tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// @module: commonjs
+// @moduleResolution: node
+// @allowJs: true
+// @traceResolution: true
+// @noEmit: true
+
+// @filename: c:/root/tsconfig.json
```
Comment (andy-ms) [inline tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts]: I would replace `c:/root/` with just `/`.

**Comment by andy-ms** [comment]
:+1:

File: tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// @filename: c:/root/typings/index.d.ts
+declare module "shortid" {
+    export var x: number;
+}
+
+// @filename: c:/root/index.ts
+/// <reference path="c:/root/typings/index.d.ts" />
+import * as foo from "shortid";
+foo.x // found in index.d.ts
+foo.y // ignored from shortid/index.ts
```
Comment (mhegazy) [inline tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts]: `shortid/index.js` and not `shortid/index.ts`.

File: tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts (hunk 1 above)
Comment (sandersn) [inline tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts]: fixed

File: tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts (hunk 2 above)
Comment (sandersn) [inline tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts]: fixed

**Comment by mhegazy** [comment]
can you add the same test with  maxNodeModuleJsDepth == 1 and then the error message should be on `x` and not `y`.

**Comment by mhegazy** [comment]
👍

**Comment by sandersn** [comment]
Actually if there is a d.ts, then it is always used. There is no way to get the compiler to prefer `node_modules` if it can find a d.ts.

File: tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts [hunk 3]
```diff
@@ -4,7 +4,7 @@
 // @traceResolution: true
 // @noEmit: true
 
-// @filename: c:/root/tsconfig.json
+// @filename: tsconfig.json
```
Comment (andy-ms) [inline tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts]: You need to include the `/` in front or else your own directory will show up in the trace.

Comment (sandersn) [inline tests/cases/compiler/maxNodeModuleJsDepthDefaultsToZero.ts]: good to know, I didn't notice that. fixed.

**Comment by sandersn** [comment]
Travis is down, but all tests pass when I merge master back into the branch, so I'll go ahead and merge. @yuit

**Pull request merged**